// connect-architect/connect.session.redis/session-ext.js
var connect = require("connect");
var RedisStore = require("connect-redis")(connect);
var assert = require("assert");

module.exports = function startup(options, imports, register) {

    assert(options.port, "option 'port' is required");
    assert(options.host, "option 'host' is required");

    var sessionStore = new RedisStore({
        port: options.port,
        host: options.host,
        pass: options.pass || "",
        prefix: options.prefix || ""
    });
    
    register(null, {
        "session-store": {
            on: sessionStore.on.bind(sessionStore),
            get: sessionStore.get.bind(sessionStore),
            set: sessionStore.set.bind(sessionStore),
            destroy: sessionStore.destroy.bind(sessionStore),
            regenerate: sessionStore.regenerate.bind(sessionStore),
            createSession: sessionStore.createSession.bind(sessionStore),
            set generate(fn) {
                sessionStore.generate = fn;
            },
            get generate() {
                return sessionStore.generate;
            }
        }
    });
};
