// architect-build/module-deps.js
var fs = require('fs');
var pathLib = require('path');

var through = require('through');

var nativeNodeModules = ["fs", "path"];

module.exports = function (mains, opts) {
    if (!Array.isArray(mains)) mains = [mains].filter(Boolean);
    
    opts = opts || Object.create(null);
    opts.modules = Object.create(null); // {id, path, contents, deps}
    
    if (!opts.ignore)
        opts.ignore = [];
    else
        opts.ignore = opts.ignore.reduce(function(map, ignore) {
            map[ignore] = true;
            return map;
        }, {});
        
    if (opts.debug)
        opts.transforms = [wrapUMD, debugSrc];

    if (!opts.transforms)
        opts.transforms = [];
    
    opts.transforms.push(removeLicenceComments, wrapUMD);
    
    if (opts.pathConfig) {
        opts.paths = opts.paths || opts.pathConfig.paths;
        opts.root = opts.root || opts.pathConfig.root;
        opts.packages = opts.packages || opts.pathConfig.packages;
    }
    
    var root = opts.root || "";
    var paths = opts.paths;
    opts.packages && opts.packages.forEach(function(pkg) {
        paths[pkg.name] = pkg;
    });
    
    function idToPath(mod, cb) {
        if (!mod.id && mod.path)
            return cb(null, mod.path);
        var id = mod.id;
        
        if (id.indexOf("!") != -1) {
            var chunks = id.split("!");
            id = chunks[1];
            mod.loaderModule = chunks[0];
        }
        
        id = resolveModuleId(id, paths);
        if (!mod.loaderModule && !/\.js$/.test(id))
            id += ".js";
        cb(null, id);
    }

    function readModule(mod, cb) {
        if (typeof mod == "string") 
            mod = {id: mod};
        
        var id = mod.id;
        if (opts.modules[id]) {
            return cb(null, null);
        }
        
        if (opts.ignore[id]) {
            ignoreDep(mod);
            return cb();
        }
        
        opts.modules[id] = mod;
        
        idToPath(mod, function(err, path) {
            var filepath = absolutePath(root, path);
            mod.path = path;
            mod.file = filepath;

            fs.readFile(filepath, 'utf8', function (err, src) {
                if (err)  {
                    if (err.code == 'ENOENT') {
                        ignoreDep(mod);
                        cb();
                        return;
                    }
                    return output.emit('error', err);
                }
                mod.source = src;
                // todo: support loaders other than text?
                if (!mod.noRequire && !mod.literal && !mod.loaderModule && !mod.noDeps) {
                    mod.deps = getDeps(src, mod.id);
                    // console.log(mod.id, mod.deps)
                    mod.deps.forEach(function(dep) {
                        loadModule(dep, mod);
                    });
                }
                cb(null, mod);
            });
        });
    }
    
    function ignoreDep(mod) {
        var parent = mod.parent || {};
        if (nativeNodeModules[mod.id] === -1) {
            console.log("file '" + mod.file + "' doesn't exisit");
            console.log("ignoring dependency '" + mod.id + "' of '" + parent.id + "'");
        }
        if (!parent.ignoredDeps)
            parent.ignoredDeps = [];
        parent.ignoredDeps.push(mod);
    }
    
    var output = through();
    
    var pending = 0;
    function loadModule(id, parent) {
        pending++;
        var mod = typeof id == "string"
            ? {id: id, parent: parent}
            : id;
            
        readModule(mod, function(err, module) {
            pending--;
            // console.log("pending", pending)
            if (module && module.source) {
                applyTransforms(module);
                output.queue(module);
            } else if (module) {
                module.parent = null;
                console.dir(module);
            }
            
            if (pending <= 0)
                end();
        });
    }
    
    function end() {
        opts._createDepMap && fs.writeFileSync("dep_map.js", JSON.stringify(opts.modules, function(key, val) {
            if (key == "source")
                return val && val[0];
            if (key == "parent")
                return val && val.id;
            return val;
        }, 4));
        process.nextTick(function() {
            output.emit("end");
        });
    }
    
    
    function applyTransforms(module) {
        if (!module.source || module.literal)
            return; // console.log(module)
        opts.transforms.forEach(function(transform) {
            transform(module);
        });
    }
    
    
    mains.forEach(function(dep) {
        loadModule(dep, {id: "#root"});
    });
    
    return output;
};


function normalizeModule(parentId, moduleName) {
    // normalize plugin requires
    if (moduleName.indexOf("!") !== -1) {
        var chunks = moduleName.split("!");
        return normalizeModule(parentId, chunks[0]) + "!" + normalizeModule(parentId, chunks[1]);
    }
    // normalize relative requires
    if (moduleName.charAt(0) == ".") {
        var base = parentId.split("/").slice(0, -1).join("/");
        moduleName = (base || parentId) + "/" + moduleName;
        
        while(moduleName.indexOf(".") !== -1 && previous != moduleName) {
            var previous = moduleName;
            moduleName = moduleName.replace(/\/\.\//, "/").replace(/[^\/]+\/\.\.\//, "");
        }
    }    
    return moduleName;
}
function getReqDeps(src, name) {
    var m = src.replace(/\s\/\/.+/g, "").match(/require\s*\(\s*(["'][^"'\n\r]+["'])\s*\)/gm);
    if (!m)
        return [];
     
    return m.map(function(r) {
        r = r.match(/["']([^"']+?)["']/)[1];
        r = normalizeModule(name, r);
        return r;
    });
}
function getAmdDeps(src, name) {
    var m = src.match(/define\(\[[^\]]+\]/gm);
    if (!m)
        return [];
     
    var deps = [];
    m.forEach(function(dep) {
        var re = /["']([^"']+?)["']/g;
        var m;
        while (m = re.exec(dep)) {
            deps.push(normalizeModule(name, m[1]));
        }
    });
    // console.log(deps);
    return deps;
}
function getDeps(src, name) {
    return getReqDeps(src, name).concat(getAmdDeps(src, name));
}
function resolveModuleId(id, paths) {
    var chunks = id.split("/");
    
    var alias = paths[chunks[0]];
    if (typeof alias == "string") {
        chunks[0] = alias;
    } else if (alias) {
        chunks[0] = alias.location;
        if (chunks.length == 1)
            chunks.push(alias.main || alias.name);
    } else if (alias === false) {
        return "";
    }
    
    return chunks.join("/");
}



function absolutePath(root, relative) {
    // console.log(root, relative)
    if (relative[0] == "/" || relative[1] == ":")
        return relative;
    return pathLib.join(root, relative);
}


// src transformations
function removeUseStrict(module) {
    module.source = module.source.replace(/['"]use strict['"];/g, "");
}

function removeLicenceComments(module) {
    module.source = module.source.replace(/(?:(;)|\n|^)\s*\/\*[\d\D]*?\*\/|\n\s*\/\/.*/g, "$1");
}
function removeLicenceCommentsKeepLines(module) {
    return module.source = module.source.replace(/(?:(;)|\n|^)\s*\/\*[\d\D]*?\*\/|\n\s*\/\/.*/g, function(cm, start) {
        return (start||"") + cm.replace(/.+/g, "");
    });
}

function wrapUMD(module) {
    if (module.loaderModule || module.noRequire)
        return;
    if (/define\(function\(/.test(module.source))
        return;
    console.log("wrapping module " + module.id);

    
    module.source = 'define(function(require, exports, module) {\n' 
        + 'var _ = {require: require, exports: exports, module: module};\n'
        + 'exports = undefined; module = undefined;\n'
        + 'function define(name, deps, m) {\n'
        + '    if (typeof name == "function") {\n'
        + '        m = name; deps = ["require", "exports", "module"]; name = _.module.id\n'
        + '    }\n'
        + '    if (typeof name == "object") {\n'
        + '        m = deps; deps = name; name = _.module.id\n'
        + '    }\n'
        + '    if (typeof deps == "function") {\n'
        + '        m = deps; deps = [];\n'
        + '    }\n'
        + '   var ret = m.apply(_.module, deps.map(function(n){return _[n] || require(n)}))\n'
        + '   if (ret) _.module.exports = ret;\n'
        + '}\n'
        + 'define.amd = true;'
        +  module.source
    + '});';
    
}

function debugSrc(module) {
    if (module.loaderModule) 
        return;
    var url = "http://localhost:8181/" + module.id;
    module.source = "try{eval(" + quote(module.source + "\n//@ sourceURL=" + url) + ");}catch(e){throw e.message + url}";
}

function quote(str) { 
    return "'"
        + str.replace(/[\\']/g, "\\$&").replace(/\n/g, "\\n")
        + "'";
}
